repair/row_level.cc

```diff
@@ -3115,3 +3117,45 @@ future<> repair_service::cleanup_history(utils::UUID repair_id) {
         rlogger.debug("repair[{}]: Finished cleaning up repair_service history", repair_id);
     });
 }
+
+future<> repair_service::load_history() {
+    auto tables = get_db().local().get_column_families();
+    for (const auto& x : tables) {
+        auto& table_uuid = x.first;
+        auto& table = x.second;
+        auto shard = unsigned(table_uuid.get_most_significant_bits()) % smp::count;
+        if (shard != this_shard_id()) {
+            continue;
+        }
+        rlogger.info("Loading repair history for keyspace={}, table={}, table_uuid={}",
+                table->schema()->ks_name(), table->schema()->cf_name(), table_uuid);
+        auto req = format("SELECT * from system.{} WHERE table_uuid = {}", db::system_keyspace::REPAIR_HISTORY, table_uuid);
+        co_await db::qctx->qp().query_internal(req, [this] (const cql3::untyped_result_set::row& row) mutable -> future<stop_iteration> {
+            auto table_uuid = row.get_as<utils::UUID>("table_uuid");
+            auto range_start = row.get_as<int64_t>("range_start");
+            auto range_end = row.get_as<int64_t>("range_end");
+            auto keyspace_name = row.get_as<sstring>("keyspace_name");
+            auto table_name = row.get_as<sstring>("table_name");
+            auto start = range_start == std::numeric_limits<int64_t>::min() ? dht::minimum_token() : dht::token::from_int64(range_start);
+            auto end = range_end == std::numeric_limits<int64_t>::min() ? dht::maximum_token() : dht::token::from_int64(range_end);
+            auto repair_time = to_gc_clock(row.get_as<db_clock::time_point>("repair_time"));
+            auto range = dht::token_range(dht::token_range::bound(start, false), dht::token_range::bound(end, true));
+            rlogger.debug("Loading repair history for keyspace={}, table={}, table_uuid={}, repair_time={}, range={}",
+                    keyspace_name, table_name, table_uuid, repair_time, range);
+            co_await get_db().invoke_on_all([table_uuid, range, repair_time, keyspace_name, table_name] (database& local_db) -> future<> {
```

Comment:
invoke_on_all() doesn't guarantee it, but since the lamba is itself in an outer coroutine, the outer coroutine should keep it alive.

Of course, debug mode may think differently, best to check.

Reply:
How about this one:

```
sharded.invoke_on_all([cap1_arg, cap2_arg] () {
    auto cap1= std::move(cap1_arg)
    auto cap2 = std::move(cap2_arg);
    co_await func1(cap1); 
    co_await func2(cap2); 
    func3(cap1); 
    co_return;
})
```

This looks much more robust since it does not depend on the outer coroutine.

Reply:
> @avikivity Could you comment on this:
> 
> sharded.invoke_on_all([cap1, cap2] () { co_await func1(cap1); // OK co_await func2(cap2); // Does the compile guarantee cap2 is still alive ? func3(cap1); // Does the compile guarantee cap1 is still alive? co_return; })

It's safe, because smp::submit_to() happens to preserve the function across until the future is resolved (unlike future::then()):

```c++
    template <typename Func>
    static futurize_t<std::result_of_t<Func()>> submit_to(unsigned t, smp_submit_to_options options, Func&& func) noexcept {
        using ret_type = std::result_of_t<Func()>;
        if (t == this_shard_id()) {
            try {
                if (!is_future<ret_type>::value) {
                   ...
                } else if (std::is_lvalue_reference<Func>::value) {
                    ...
                } else {
                    // Deferring call on rvalue function, make sure to preserve it across call
                    auto w = std::make_unique<std::decay_t<Func>>(std::move(func));
                    auto ret = futurize<ret_type>::invoke(*w);
                    return ret.finally([w = std::move(w)] {});
                }
            } catch (...) {
                  ...
            }
        } else {
            return _qs[t][this_shard_id()].submit(t, options, std::forward<Func>(func));
        }
    }
```

Reply:
OK. Thanks.